sympy/algebras/quaternion.py

```diff
@@ -0,0 +1,534 @@
+# References :
+# http://www.euclideanspace.com/maths/algebra/realNormedAlgebra/quaternions/
+# https://en.wikipedia.org/wiki/Quaternion
+
+from __future__ import print_function
+
+from sympy.core.expr import Expr
+from sympy import Rational
+from sympy import re, im, conjugate
+from sympy import sqrt, sin, cos, acos, asin, exp, ln
+from sympy import trigsimp
+from sympy import diff, integrate
+from sympy import Matrix, Add, Mul
+from sympy import symbols, sympify
+from sympy.printing.latex import latex
+from sympy.printing import StrPrinter
+
+
+class Quaternion(Expr):
+    """Provides basic quaternion operations.
+    Quaternion objects can be instantiated as Quaternion(a, b, c, d)
+    as in (a + b*i + c*j + d*k).
+
+    Example
+    ========
+
+    >>> from sympy.algebras.quaternion import Quaternion
+    >>> q = Quaternion(1, 2, 3, 4)
+    >>> q
+    1 + 2*i + 3*j + 4*k
+    """
+    _op_priority = 11.0
+
+    is_commutative = False
+
+    def __new__(cls, a=0, b=0, c=0, d=0):
+        a = sympify(a)
+        b = sympify(b)
+        c = sympify(c)
+        d = sympify(d)
+
+        if any(i.is_real is False for i in [a, b, c, d]):
```

Comment:
I'm not sure that checking for reality is necessary. It should be enough that the coordinates are commutative. It makes perfect sense to define a quaternion algebra over any commutative ring, only it need not be a (skew) field in general.

Reply:
@jksuom I see your point. I'll change 
```is_real``` to ```is_commutative```.